accord-core/src/main/java/accord/utils/async/AsyncResults.java

```diff
@@ -0,0 +1,309 @@
+            this.failure = failure;
+        }
+
+        @Override
+        public void addCallback(BiConsumer<? super V, Throwable> callback)
+        {
+            callback.accept(value, failure);
+        }
+
+        @Override
+        public boolean isDone()
+        {
+            return true;
+        }
+
+        @Override
+        public boolean isSuccess()
+        {
+            return failure == null;
+        }
+    }
+
+    /**
+     * Creates an AsyncResult for the given chain. This calls begin on the supplied chain
+     */
+    public static <V> AsyncResult<V> forChain(AsyncChain<V> chain)
+    {
+        return new Chain<>(chain);
+    }
+
+    public static <V> AsyncResult<V> success(V value)
+    {
+        return new Immediate<>(value);
+    }
+
+    public static <V> AsyncResult<V> failure(Throwable failure)
+    {
+        return new Immediate<>(failure);
+    }
+
+    public static <V> AsyncResult<V> ofCallable(Executor executor, Callable<V> callable)
+    {
+        Settable<V> result = new Settable<V>();
+        executor.execute(() -> {
+            try
+            {
+                result.trySuccess(callable.call());
+            }
+            catch (Exception e)
+            {
+                result.tryFailure(e);
+            }
+        });
+        return result;
+    }
+
+    public static AsyncResult<Void> ofRunnable(Executor executor, Runnable runnable)
+    {
+        Settable<Void> result = new Settable<Void>();
+        executor.execute(() -> {
+            try
+            {
+                runnable.run();
+                result.trySuccess(null);
+            }
+            catch (Exception e)
+            {
+                result.tryFailure(e);
+            }
+        });
+        return result;
+    }
+
+    public static <V> AsyncResult.Settable<V> settable()
+    {
+        return new Settable<>();
+    }
+
+    private static <V> List<AsyncChain<V>> toChains(List<AsyncResult<V>> results)
+    {
+        List<AsyncChain<V>> chains = new ArrayList<>(results.size());
+        for (int i=0,mi=results.size(); i<mi; i++)
+            chains.add(results.get(i).toChain());
+        return chains;
+    }
+
+    public static <V> AsyncChain<List<V>> all(List<AsyncResult<V>> results)
+    {
+        Preconditions.checkArgument(!results.isEmpty());
+        return new AsyncChainCombiner.All<>(toChains(results));
+    }
+
+    public static <V> AsyncChain<V> reduce(List<AsyncResult<V>> results, BiFunction<V, V, V> reducer)
+    {
+        Preconditions.checkArgument(!results.isEmpty());
+        if (results.size() == 1)
+            return results.get(0).toChain();
+        return new AsyncChainCombiner.Reduce<>(toChains(results), reducer);
+    }
+
+    public static <V> V getBlocking(AsyncResult<V> asyncResult) throws InterruptedException, ExecutionException
+    {
+        AtomicReference<Result<V>> callbackResult = new AtomicReference<>();
+        CountDownLatch latch = new CountDownLatch(1);
```

Comment:
Hmm. We have a problem here which is that these aren't very compatible with the Simulator. We are meant to use `newCountDownLatch` in Cassandra for this.

Reply:
Now we use byte weaving we have a lot of power to take over things we didn’t initially. So yes in principle we can capture these. There’s two ways, 
1) by providing an overriding implementation and replacing allocations, as we do for ConcurrentHashMap - this isn’t ideal though and I think doesn’t work for all juc classes as I think not all can be fully overridden
2) via LockSupport intercept

(2) should theoretically already be supported, but is inefficient and not used extensively so might not be fully working - this patch does break simulation though, and I haven’t had time to fully dig into all of the issues, but I might actually be wrong about this being one of the reasons. I will have to dig in further.